Error injecting constructor with Unrecognized Type: [null]

CreationException: Unable to create injector, see the following errors:

```
1) Error injecting constructor, java.lang.IllegalArgumentException: Unrecognized Type: [null]
  at play.modules.swagger.SwaggerPluginImpl.<init>(SwaggerPlugin.scala:33)
  while locating play.modules.swagger.SwaggerPluginImpl
  at play.modules.swagger.SwaggerModule.bindings(SwaggerModule.scala:11):
Binding(interface play.modules.swagger.SwaggerPlugin to ConstructionTarget(class play.modules.swagger.SwaggerPluginImpl) eagerly) (via modules: com.google.inject.util.Modules$OverrideModule -> play.api.inject.guice.GuiceableModuleConversions$$anon$1)
  while locating play.modules.swagger.SwaggerPlugin

1 error
        at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:466) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:184) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:110) ~[guice-4.0.jar:na]
        at com.google.inject.Guice.createInjector(Guice.java:96) ~[guice-4.0.jar:na]
        at com.google.inject.Guice.createInjector(Guice.java:73) ~[guice-4.0.jar:na]
        at com.google.inject.Guice.createInjector(Guice.java:62) ~[guice-4.0.jar:na]
        at play.api.inject.guice.GuiceBuilder.injector(GuiceInjectorBuilder.scala:126) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:93) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.api.inject.guice.GuiceApplicationLoader.load(GuiceApplicationLoader.scala:21) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1$$anonfun$2.apply(DevServerStart.scala:153) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1$$anonfun$2.apply(DevServerStart.scala:150) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.utils.Threads$.withContextClassLoader(Threads.scala:21) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:150) ~[play-server_2.11-2.4.6.jar:2.4.6]
        ... 14 common frames omitted
Caused by: java.lang.IllegalArgumentException: Unrecognized Type: [null]
        at com.fasterxml.jackson.databind.type.TypeFactory._constructType(TypeFactory.java:405) ~[jackson-databind-2.5.4.jar:2.5.4]
        at com.fasterxml.jackson.databind.type.TypeFactory.constructType(TypeFactory.java:354) ~[jackson-databind-2.5.4.jar:2.5.4]
        at com.fasterxml.jackson.databind.ObjectMapper.constructType(ObjectMapper.java:1332) ~[jackson-databind-2.5.4.jar:2.5.4]
        at io.swagger.scala.converter.SwaggerScalaModelConverter.resolveProperty(SwaggerScalaModelConverter.scala:32) ~[swagger-scala-module_2.11-1.0.0.jar:1.0.0]
        at io.swagger.converter.ModelConverterContextImpl.resolveProperty(ModelConverterContextImpl.java:79) ~[swagger-core-1.5.4.jar:1.5.4]
        at io.swagger.converter.ModelConverters.readAsProperty(ModelConverters.java:58) ~[swagger-core-1.5.4.jar:1.5.4]
        at play.modules.swagger.PlayReader.createProperty(PlayReader.java:644) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.PlayReader.getParameters(PlayReader.java:569) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.PlayReader.parseMethod(PlayReader.java:514) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.PlayReader.read(PlayReader.java:136) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.PlayReader.read(PlayReader.java:60) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.PlayReader.read(PlayReader.java:54) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.ApiListingCache$$anonfun$listing$1.apply(ApiListingCache.scala:17) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.ApiListingCache$$anonfun$listing$1.apply(ApiListingCache.scala:11) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at scala.Option.orElse(Option.scala:289) ~[scala-library-2.11.7.jar:na]
        at play.modules.swagger.ApiListingCache$.listing(ApiListingCache.scala:11) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.SwaggerPluginImpl.<init>(SwaggerPlugin.scala:144) ~[swagger-play2_2.11-1.5.0.jar:1.5.0]
        at play.modules.swagger.SwaggerPluginImpl$$FastClassByGuice$$de7219b8.newInstance(<generated>) ~[guice-4.0.jar:1.5.0]
        at com.google.inject.internal.cglib.reflect.$FastConstructor.newInstance(FastConstructor.java:40) ~[guice-4.0.jar:na]
        at com.google.inject.internal.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:61) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:105) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:85) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:267) ~[guice-4.0.jar:na]
        at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:56) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1103) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40) ~[guice-4.0.jar:na]
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:145) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:205) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:199) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1092) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:199) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:180) ~[guice-4.0.jar:na]
        ... 25 common frames omitted
```

The stacktrace leads to an error in the resolveProperty method in the SwaggerScalaModelConverter.scala:32. Causing constructType to fail in

    val javaType = Json.mapper().constructType(`type`)

So it seems that resolveProperty is the culprit of the issue otherwise (if I misinterpret the stacktrace) a generic error with dependency injection might be also a candidate. 

In SwaggerPlugin.scala:33 the dependency injection is done:

    class SwaggerPluginImpl @Inject()(lifecycle: ApplicationLifecycle, router: Router, app: Application) extends SwaggerPlugin {


With that very generic failure any help would be greatly appreciated.

I'm still seeing the same issue after upgrading to swagger-1.5.1 (with
swagger-parser-1.0.16).
None of my models are annotated (only the controllers are).

Unfortunatly the stacktrace does not reveal any hints on where to look in
my models/controllers.

```
Error injecting constructor, java.lang.IllegalArgumentException:
Unrecognized Type: [null]
  at play.modules.swagger.SwaggerPluginImpl.<init>(SwaggerPlugin.scala:33)
  while locating play.modules.swagger.SwaggerPluginImpl
  at play.modules.swagger.SwaggerModule.bindings(SwaggerModule.scala:11):
Binding(interface play.modules.swagger.SwaggerPlugin to
ConstructionTarget(class play.modules.swagger.SwaggerPluginImpl) eagerly)
(via modules: com.google.inject.util.Modules$OverrideModule ->
play.api.inject.guice.GuiceableModuleConversions$$anon$1)
  while locating play.modules.swagger.SwaggerPlugin

1 error
at
com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:466)
~[guice-4.0.jar:na]
at
com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:184)
~[guice-4.0.jar:na]
at
com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:110)
~[guice-4.0.jar:na]
at com.google.inject.Guice.createInjector(Guice.java:96) ~[guice-4.0.jar:na]
at com.google.inject.Guice.createInjector(Guice.java:73) ~[guice-4.0.jar:na]
Caused by: java.lang.IllegalArgumentException: Unrecognized Type: [null]
at
com.fasterxml.jackson.databind.type.TypeFactory._constructType(TypeFactory.java:405)
~[jackson-databind-2.5.4.jar:2.5.4]
at
com.fasterxml.jackson.databind.type.TypeFactory.constructType(TypeFactory.java:354)
~[jackson-databind-2.5.4.jar:2.5.4]
at
com.fasterxml.jackson.databind.ObjectMapper.constructType(ObjectMapper.java:1332)
~[jackson-databind-2.5.4.jar:2.5.4]
at
io.swagger.scala.converter.SwaggerScalaModelConverter.resolveProperty(SwaggerScalaModelConverter.scala:32)
~[swagger-scala-module_2.11-1.0.0.jar:1.0.0]
at
io.swagger.converter.ModelConverterContextImpl.resolveProperty(ModelConverterContextImpl.java:79)
~[swagger-core-1.5.6.jar:1.5.6]
```

On Fri, Jan 15, 2016 at 10:48 AM, Rafał Gajdulewicz <
notifications@github.com> wrote:

> My issue was fixed in version 1.5.1, thanks!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/swagger-api/swagger-play/issues/43#issuecomment-171915463
> .

**EDIT:** Never mind my issue, it was induced by swagger but is not related, I had a flag that was being initialized in the globalsettings section that starts my bindings which relied on play to be running by then and was thus failing, after eliminating that it was smooth sailing.

**OP:**

Seems like I'm in the same boat, using v1.5.1 this is my stack trace:

```
play.api.UnexpectedException: Unexpected exception[CreationException: Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.ExceptionInInitializerError
  at play.modules.swagger.SwaggerPluginImpl.<init>(SwaggerPlugin.scala:33)
  while locating play.modules.swagger.SwaggerPluginImpl
  at play.modules.swagger.SwaggerModule.bindings(SwaggerModule.scala:11):
Binding(interface play.modules.swagger.SwaggerPlugin to ConstructionTarget(class play.modules.swagger.SwaggerPluginImpl) eagerly) (via modules: com.google.inject.util.Modules$OverrideModule -> play.api.inject.guice.GuiceableModuleConversions$$anon$1)
  while locating play.modules.swagger.SwaggerPlugin

1 error]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:165) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:121) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at scala.Option.map(Option.scala:146) ~[scala-library-2.11.6.jar:na]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1.apply(DevServerStart.scala:121) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1.apply(DevServerStart.scala:119) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at scala.util.Success.flatMap(Try.scala:230) ~[scala-library-2.11.6.jar:na]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1.apply(DevServerStart.scala:119) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1.apply(DevServerStart.scala:111) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24) ~[scala-library-2.11.6.jar:na]
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24) ~[scala-library-2.11.6.jar:na]
        at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402) ~[na:1.8.0_60]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[na:1.8.0_60]
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[na:1.8.0_60]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[na:1.8.0_60]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[na:1.8.0_60]
Caused by: com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.ExceptionInInitializerError
  at play.modules.swagger.SwaggerPluginImpl.<init>(SwaggerPlugin.scala:33)
  while locating play.modules.swagger.SwaggerPluginImpl
  at play.modules.swagger.SwaggerModule.bindings(SwaggerModule.scala:11):
Binding(interface play.modules.swagger.SwaggerPlugin to ConstructionTarget(class play.modules.swagger.SwaggerPluginImpl) eagerly) (via modules: com.google.inject.util.Modules$OverrideModule -> play.api.inject.guice.GuiceableModuleConversions$$anon$1)
  while locating play.modules.swagger.SwaggerPlugin

1 error
        at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:466) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:184) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:110) ~[guice-4.0.jar:na]
        at com.google.inject.Guice.createInjector(Guice.java:96) ~[guice-4.0.jar:na]
        at com.google.inject.Guice.createInjector(Guice.java:73) ~[guice-4.0.jar:na]
        at com.google.inject.Guice.createInjector(Guice.java:62) ~[guice-4.0.jar:na]
        at play.api.inject.guice.GuiceBuilder.injector(GuiceInjectorBuilder.scala:126) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:93) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.api.inject.guice.GuiceApplicationLoader.load(GuiceApplicationLoader.scala:21) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1$$anonfun$2.apply(DevServerStart.scala:153) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1$$anonfun$2.apply(DevServerStart.scala:150) ~[play-server_2.11-2.4.6.jar:2.4.6]
        at play.utils.Threads$.withContextClassLoader(Threads.scala:21) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:150) ~[play-server_2.11-2.4.6.jar:2.4.6]
        ... 14 common frames omitted
Caused by: java.lang.ExceptionInInitializerError: null
        at java.lang.Class.forName0(Native Method) ~[na:1.8.0_60]
        at java.lang.Class.forName(Class.java:348) ~[na:1.8.0_60]
        at play.modules.swagger.util.SwaggerContext$$anonfun$loadClass$1.apply(SwaggerContext.scala:27) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.util.SwaggerContext$$anonfun$loadClass$1.apply(SwaggerContext.scala:24) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at scala.collection.immutable.List.foreach(List.scala:381) ~[scala-library-2.11.6.jar:na]
        at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35) ~[scala-library-2.11.6.jar:na]
        at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45) ~[scala-library-2.11.6.jar:na]
        at play.modules.swagger.util.SwaggerContext$.loadClass(SwaggerContext.scala:24) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.PlayApiScanner$$anonfun$1.applyOrElse(PlayApiScanner.scala:86) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.PlayApiScanner$$anonfun$1.applyOrElse(PlayApiScanner.scala:83) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at scala.collection.immutable.List.collect(List.scala:295) ~[scala-library-2.11.6.jar:na]
        at play.modules.swagger.PlayApiScanner.classes(PlayApiScanner.scala:83) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.ApiListingCache$$anonfun$listing$1.apply(ApiListingCache.scala:15) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.ApiListingCache$$anonfun$listing$1.apply(ApiListingCache.scala:11) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at scala.Option.orElse(Option.scala:289) ~[scala-library-2.11.6.jar:na]
        at play.modules.swagger.ApiListingCache$.listing(ApiListingCache.scala:11) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.SwaggerPluginImpl.<init>(SwaggerPlugin.scala:160) ~[swagger-play2_2.11-1.5.1.jar:1.5.1]
        at play.modules.swagger.SwaggerPluginImpl$$FastClassByGuice$$de7219b8.newInstance(<generated>) ~[guice-4.0.jar:1.5.1]
        at com.google.inject.internal.cglib.reflect.$FastConstructor.newInstance(FastConstructor.java:40) ~[guice-4.0.jar:na]
        at com.google.inject.internal.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:61) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:105) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:85) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:267) ~[guice-4.0.jar:na]
        at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:56) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1103) ~[guice-4.0.jar:na]
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40) ~[guice-4.0.jar:na]
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:145) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:205) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:199) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1092) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:199) ~[guice-4.0.jar:na]
        at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:180) ~[guice-4.0.jar:na]
        ... 25 common frames omitted
Caused by: java.lang.RuntimeException: There is no started application
        at scala.sys.package$.error(package.scala:27) ~[scala-library-2.11.6.jar:na]
        at play.api.Play$$anonfun$current$1.apply(Play.scala:71) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.api.Play$$anonfun$current$1.apply(Play.scala:71) ~[play_2.11-2.4.6.jar:2.4.6]
        at scala.Option.getOrElse(Option.scala:121) ~[scala-library-2.11.6.jar:na]
        at play.api.Play$.current(Play.scala:71) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.api.Play.current(Play.scala) ~[play_2.11-2.4.6.jar:2.4.6]
        at play.Play.isDev(Play.java:31) ~[play_2.11-2.4.6.jar:2.4.6]
        at controllers.core.RestController.<clinit>(RestController.java:40) ~[na:na]
        ... 59 common frames omitted
```

I've only used swagger on a single location:
`@ApiOperation(value = "Get Companies", notes = "", httpMethod = "GET")`
for the following routes:

```
GET            /api/company/                                @controllers.company.CompanyController.getCompany(id: Int ?= -1)
GET            /api/company/:id                             @controllers.company.CompanyController.getCompany(id: Int)
```

--- a/play-2.4/swagger-play2/app/play/modules/swagger/PlayReader.java
+++ b/play-2.4/swagger-play2/app/play/modules/swagger/PlayReader.java
@@ -1,7 +1,6 @@
 package play.modules.swagger;
 
 import com.fasterxml.jackson.databind.JavaType;
-import com.fasterxml.jackson.databind.type.TypeFactory;
 import io.swagger.annotations.*;
 import io.swagger.annotations.Info;
 import io.swagger.converter.ModelConverters;
@@ -13,6 +12,7 @@ import io.swagger.models.parameters.*;
 import io.swagger.models.parameters.Parameter;
 import io.swagger.models.properties.*;
 import io.swagger.util.BaseReaderUtils;
+import io.swagger.util.Json;
 import io.swagger.util.ParameterProcessor;
 import io.swagger.util.PrimitiveType;
 import io.swagger.util.ReflectionUtils;
@@ -20,11 +20,14 @@ import org.apache.commons.lang3.StringUtils;
 import play.Logger;
 import play.modules.swagger.util.CrossUtil;
 import play.routes.compiler.*;
+import scala.Option;
 
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Method;
 import java.lang.reflect.Type;
 import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public class PlayReader {
 
@@ -192,7 +195,7 @@ public class PlayReader {
                         }
                         path.set(httpMethod, operation);
                         try {
-                            readImplicitParameters(method, operation);
+                            readImplicitParameters(method, operation, cls);
                         } catch (Exception e) {
                             throw e;
                         }
@@ -350,11 +353,11 @@ public class PlayReader {
         info.getVendorExtensions().putAll(BaseReaderUtils.parseExtensions(infoConfig.extensions()));
     }
 
-    private void readImplicitParameters(Method method, Operation operation) {
+    private void readImplicitParameters(Method method, Operation operation, Class<?> cls) {
         ApiImplicitParams implicitParams = method.getAnnotation(ApiImplicitParams.class);
         if (implicitParams != null && implicitParams.value().length > 0) {
             for (ApiImplicitParam param : implicitParams.value()) {
-                Parameter p = readImplicitParam(param);
+                Parameter p = readImplicitParam(param, cls);
                 if (p != null) {
                     operation.addParameter(p);
                 }
@@ -362,7 +365,7 @@ public class PlayReader {
         }
     }
 
-    protected io.swagger.models.parameters.Parameter readImplicitParam(ApiImplicitParam param) {
+    protected io.swagger.models.parameters.Parameter readImplicitParam(ApiImplicitParam param, Class<?> cls) {
         final Parameter p;
         if (param.paramType().equalsIgnoreCase("path")) {
             p = new PathParameter();
@@ -380,18 +383,31 @@ public class PlayReader {
         }
         Type type = null;
         // Swagger ReflectionUtils can't handle file or array datatype
-        if (!"file".equalsIgnoreCase(param.dataType()) && !"array".equalsIgnoreCase(param.dataType())){
-            type = typeFromString(param.dataType());
+        if (!"".equalsIgnoreCase(param.dataType()) && !"file".equalsIgnoreCase(param.dataType()) && !"array".equalsIgnoreCase(param.dataType())){
+            type = typeFromString(param.dataType(), cls);
+
         }
-        return ParameterProcessor.applyAnnotations(getSwagger(), p, type == null ? String.class : type, Collections.singletonList(param));
+        Parameter result =  ParameterProcessor.applyAnnotations(getSwagger(), p, type == null ? String.class : type, Collections.singletonList(param));
+
+        if (result instanceof AbstractSerializableParameter && type != null) {
+            Property schema = createProperty(type);
+            ((AbstractSerializableParameter)p).setProperty(schema);
+        }
+
+        return result;
+
     }
 
-    private static Type typeFromString(String type) {
+    private static Type typeFromString(String type, Class<?> cls) {
         final PrimitiveType primitive = PrimitiveType.fromName(type);
         if (primitive != null) {
             return primitive.getKeyClass();
         }
         try {
+            Type routeType = getOptionTypeFromString (type, cls);
+
+            if (routeType != null) return routeType;
+
             return Thread.currentThread().getContextClassLoader().loadClass(type);
         } catch (Exception e) {
             Logger.error(String.format("Failed to resolve '%s' into class", type), e);
@@ -522,37 +538,69 @@ public class PlayReader {
         return operation;
     }
 
-    private Type getParamType(Class<?> cls, Method method, String simpleTypeName) {
-        Type[] genericParameterTypes = method.getGenericParameterTypes();
-        for (Type genericParameterType : genericParameterTypes) {
-            final Type type = TypeFactory.defaultInstance().constructType(genericParameterType, cls);
-            final Class<?> paramClass = ((JavaType) type).getRawClass();
-            if (simpleTypeName.equalsIgnoreCase(paramClass.getSimpleName())) {
-                return type;
-            }
-            if (simpleTypeName.equalsIgnoreCase(paramClass.getName())) {
-                return type;
+
+    final static class OptionTypeResolver {
+        private Option<Integer> optionTypeInt;
+        private Option<Long> optionTypeLong;
+        private Option<Byte> optionTypeByte;
+        private Option<Boolean> optionTypeBoolean;
+        private Option<Character> optionTypeChar;
+        private Option<Float> optionTypeFloat;
+        private Option<Double> optionTypeDouble;
+        private Option<Short> optionTypeShort;
+
+        static Type resolveOptionType (String innerType, Class<?> cls) {
+            try {
+                return Json.mapper().getTypeFactory().constructType(
+                        OptionTypeResolver.class.getDeclaredField("optionType" + innerType).getGenericType(), cls);
+            } catch (NoSuchFieldException e) {
+                return null;
             }
         }
-        return null;
+    }
+
+    private static Type getOptionTypeFromString (String simpleTypeName, Class<?> cls) {
+
+        if (simpleTypeName == null) return null;
+        String regex = "(Option|scala\\.Option)\\s*\\[\\s*(Int|Long|Float|Double|Byte|Short|Char|Boolean)\\s*\\]\\s*$";
+
+        Pattern pattern = Pattern.compile(regex);
+        Matcher matcher = pattern.matcher(simpleTypeName);
+        if (matcher.find())
+        {
+            String enhancedType = matcher.group(2);
+            return OptionTypeResolver.resolveOptionType(enhancedType, cls);
+        } else {
+            return null;
+        }
+    }
+    private Type getParamType(Class<?> cls, Method method, String simpleTypeName, int position) {
+
+        try {
+            Type type = getOptionTypeFromString (simpleTypeName, cls);
+            if (type != null) return type;
+
+            Type[] genericParameterTypes = method.getGenericParameterTypes();
+            return Json.mapper().getTypeFactory().constructType(genericParameterTypes[position], cls);
+        } catch (Exception e) {
+            Logger.error(String.format("Exception getting parameter type for method %s, param %s at position %d"), e);
+            return null;
+        }
+
     }
 
     private List<Annotation> getParamAnnotations(Class<?> cls, Type[] genericParameterTypes, Annotation[][] paramAnnotations, String simpleTypeName, int fieldPosition) {
-        Type type = TypeFactory.defaultInstance().constructType(genericParameterTypes[fieldPosition], cls);
-        Class<?> paramClass = ((JavaType)type).getRawClass();
-        if (simpleTypeName.equalsIgnoreCase(paramClass.getSimpleName())) {
+        try {
             return Arrays.asList(paramAnnotations[fieldPosition]);
+        } catch (Exception e) {
+            Logger.error(String.format("Exception getting parameter type for method %s, param %s at position %d"), e);
+            return null;
         }
-        if (simpleTypeName.equalsIgnoreCase(paramClass.getName())) {
-            return Arrays.asList(paramAnnotations[fieldPosition]);
-        }
-        return null;
     }
 
     private List<Annotation> getParamAnnotations(Class<?> cls, Method method, String simpleTypeName, int fieldPosition) {
         Type[] genericParameterTypes = method.getGenericParameterTypes();
         Annotation[][] paramAnnotations = method.getParameterAnnotations();
-
         List<Annotation> annotations = getParamAnnotations(cls, genericParameterTypes, paramAnnotations, simpleTypeName, fieldPosition);
         if (annotations != null) {
             return annotations;
@@ -587,7 +635,7 @@ public class PlayReader {
             if (def.startsWith("\"") && def.endsWith("\"")){
                 def = def.substring(1,def.length()-1);
             }
-            Type type = getParamType(cls, method, p.typeName());
+            Type type = getParamType(cls, method, p.typeName(), fieldPosition);
             Property schema = createProperty(type);
             if (route.path().has(p.name())) {
                 // it's a path param
@@ -602,9 +650,7 @@ public class PlayReader {
             }
             parameter.setName(p.name());
             List<Annotation> annotations = getParamAnnotations(cls, method, p.typeName(), fieldPosition);
-
             ParameterProcessor.applyAnnotations(getSwagger(), parameter, type, annotations);
-
             parameters.add(parameter);
             fieldPosition++;
         }
@@ -623,7 +669,7 @@ public class PlayReader {
     }
 
     private static boolean isVoid(Type type) {
-        final Class<?> cls = TypeFactory.defaultInstance().constructType(type).getRawClass();
+        final Class<?> cls = Json.mapper().getTypeFactory().constructType(type).getRawClass();
         return Void.class.isAssignableFrom(cls) || Void.TYPE.isAssignableFrom(cls);
     }
 
@@ -631,7 +677,7 @@ public class PlayReader {
         if (type == null) {
             return false;
         }
-        final JavaType javaType = TypeFactory.defaultInstance().constructType(type);
+        final JavaType javaType = Json.mapper().getTypeFactory().constructType(type);
         if (isVoid(javaType)) {
             return false;
         }

--- a/play-2.4/swagger-play2/build.sbt
+++ b/play-2.4/swagger-play2/build.sbt
@@ -8,8 +8,8 @@ crossScalaVersions := Seq("2.11.6", "2.11.7")
 
 libraryDependencies ++= Seq(
   "org.slf4j"          % "slf4j-api"                  % "1.6.4",
-  "io.swagger"         % "swagger-core"               % "1.5.7",
-  "io.swagger"        %% "swagger-scala-module"       % "1.0.0",
+  "io.swagger"         % "swagger-core"               % "1.5.8-SNAPSHOT",
+  "io.swagger"        %% "swagger-scala-module"       % "1.0.2-SNAPSHOT",
   "com.typesafe.play" %% "routes-compiler"            % "2.4.6",
   "com.typesafe.play" %% "play-ebean"                 % "2.0.0"            % "test",
   "org.specs2"        %% "specs2-core"                % "3.6.6"            % "test",

--- a/play-2.4/swagger-play2/test/PlayApiListingCacheSpec.scala
+++ b/play-2.4/swagger-play2/test/PlayApiListingCacheSpec.scala
@@ -2,7 +2,7 @@ import java.io.File
 
 import io.swagger.config.ScannerFactory
 import io.swagger.models.{ModelImpl, HttpMethod}
-import io.swagger.models.parameters.{BodyParameter, PathParameter}
+import io.swagger.models.parameters.{QueryParameter, BodyParameter, PathParameter}
 import io.swagger.models.properties.{RefProperty, ArrayProperty}
 import play.modules.swagger._
 import org.specs2.mutable._
@@ -24,6 +24,7 @@ GET /api/pointsofinterest testdata.PointOfInterestController.list(eastingMin:Dou
 GET /api/dog testdata.DogController.list
 PUT /api/dog testdata.DogController.add1
 GET /api/cat @testdata.CatController.list
+GET /api/cat43 @testdata.CatController.testIssue43(test_issue_43_param: Option[Int])
 PUT /api/cat @testdata.CatController.add1
 GET /api/fly testdata.FlyController.list
 PUT /api/dog testdata.DogController.add1
@@ -72,11 +73,13 @@ PUT /api/dog/:id testdata.DogController.add0(id:String)
       val docRoot = ""
       val swagger = ApiListingCache.listing(docRoot, "127.0.0.1")
 
-      Logger.debug ("swagger: " + toJsonString(swagger.get))
+      Logger.debug ("swagger: " + toJsonString(swagger))
+      swagger must beSome
+
       swagger must beSome
       swagger.get.getSwagger must beEqualTo("2.0")
       swagger.get.getBasePath must beEqualTo(basePath)
-      swagger.get.getPaths.size must beEqualTo(6)
+      swagger.get.getPaths.size must beEqualTo(7)
       swagger.get.getDefinitions.size must beEqualTo(5)
       swagger.get.getHost must beEqualTo(swaggerConfig.getHost)
       swagger.get.getInfo.getContact.getName must beEqualTo(swaggerConfig.getContact)
@@ -132,6 +135,23 @@ PUT /api/dog/:id testdata.DogController.add0(id:String)
       opCatPut.getConsumes must beNull
       opCatPut.getResponses.get("200").getSchema.asInstanceOf[RefProperty].getSimpleRef must beEqualTo("ActionAnyContent")
       opCatPut.getProduces must beNull
+
+      val pathCat43 = swagger.get.getPaths.get("/cat43")
+      pathCat43.getOperations.size must beEqualTo(1)
+
+      val opCatGet43 = pathCat43.getOperationMap.get(HttpMethod.GET)
+      opCatGet43.getOperationId must beEqualTo("test issue #43_nick")
+      opCatGet43.getResponses.get("200").getSchema.asInstanceOf[ArrayProperty].getItems.asInstanceOf[RefProperty].getSimpleRef must beEqualTo("Cat")
+
+      opCatGet43.getParameters.head.getName must beEqualTo("test_issue_43_param")
+      opCatGet43.getParameters.head.getIn must beEqualTo("query")
+      opCatGet43.getParameters.head.asInstanceOf[QueryParameter].getType must beEqualTo("integer")
+
+      opCatGet43.getParameters.get(1).getName must beEqualTo("test_issue_43_implicit_param")
+      opCatGet43.getParameters.get(1).getIn must beEqualTo("query")
+      opCatGet43.getParameters.get(1).asInstanceOf[QueryParameter].getType must beEqualTo("integer")
+
+
 
       val pathDog = swagger.get.getPaths.get("/dog")
       pathDog.getOperations.size must beEqualTo(2)

--- a/play-2.4/swagger-play2/test/testdata/CatController.scala
+++ b/play-2.4/swagger-play2/test/testdata/CatController.scala
@@ -55,6 +55,18 @@ class CatController extends Controller {
     def no_route = Action {
       request => Ok("test case")
     }
+
+  @ApiOperation(value = "test issue #43",
+    nickname = "test issue #43_nick",
+    notes = "test issue #43_notes",
+    response = classOf[testdata.Cat],
+    responseContainer = "List",
+    httpMethod = "GET")
+  @ApiImplicitParams(Array(
+    new ApiImplicitParam(name = "test_issue_43_implicit_param", dataType = "Option[Int]", value = "test issue #43 implicit param", paramType = "query")))
+  def testIssue43(test_issue_43_param:  Option[Int]) = Action {
+    request => Ok("test issue #43")
+  }
 }
 
 case class Cat(id: Long, name: String)

